File: src/algorithms/cpp/Longest_Increasing_Path_in_a_Matrix.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}


class Solution {
private:
    int moves[4][2] = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    int m, n;
    vector <vector<int>> rec, matrix;
    map <pair<int, int>, int> path;
    

    void dfs(int step, int x, int y) {
        rec[x][y] = step;
        for (int i = 0; i < 4; ++ i) {
            int nx = x + moves[i][0], ny = y + moves[i][1];
            if (nx >= 0 && nx < m && ny >= 0 && ny < n && matrix[nx][ny] > matrix[x][y] && step + 1 > rec[nx][ny]) {
                dfs(step + 1, nx, ny);
            }
        }
    }


public:
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        if (SZ(matrix) == 0)    return 0;
        this->matrix = matrix;
        m = SZ(matrix), n = SZ(matrix[0]);
        for (int i = 0; i < m; ++ i) {
            vector <int> tmp(n, 0);
            rec.push_back(tmp);
        }
        for (int i = 0; i < m; ++ i)
            for (int j = 0; j < n; ++ j)
                if (rec[i][j] == 0)
                    dfs(1, i, j);
        int result = 0;
        for (int i = 0; i < m; ++ i) {
            for (int j = 0; j < n; ++ j) {
                get_max(result, rec[i][j]);
            }
        }
        return result;
    }
};


int main() {
    Solution solution;
    vector <vector<int>> mp = {{9, 9, 4}, {6, 6, 8}, {2, 1, 1}};
    cout << solution.longestIncreasingPath(mp) << endl;
    return 0;
}

File: src/algorithms/python/Longest_Increasing_Path_in_a_Matrix.py
#!/bin/env python2
# -*- coding: utf-8 -* -


class Solution(object):
    def dfs(self, x, y, now):
        moves = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        for x_diff, y_diff in moves:
            _x, _y = x + x_diff, y + y_diff
            if (
                0 <= _x < self.m and
                0 <= _y < self.n and
                self.matrix[_x][_y] > self.matrix[x][y] and
                self.rec[_x][_y] < now + 1
            ):
                self.rec[_x][_y] = now + 1
                self.dfs(_x, _y, now + 1)

    def longestIncreasingPath(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: int
        """
        if len(matrix) == 0:
            return 0
        self.matrix = matrix
        self.m, self.n = len(matrix), len(matrix[0])
        self.rec = [[0 for i in range(self.n)] for j in range(self.m)]
        for x in range(self.m):
            for y in range(self.n):
                if self.rec[x][y] <= 1:
                    self.rec[x][y] = 1
                    self.dfs(x, y, 1)
        return max(map(max, self.rec))
